### PR TITLE
Populate transport_failure_reason when a stream is reset due to QUIC connection close.

### PR DESCRIPTION
**Pull request by RenjieTang**
Commit Message: Populate transport_failure_reason when a stream is reset due to QUIC connection close.
Additional Description: Currently we only populate for connection closes. We can consider reporting error details for stream reset with errors as well.
Risk Level: low
Testing: existing tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: source/common/quic/envoy_quic_client_stream.cc [hunk 1]
```diff
@@ -354,15 +355,16 @@ void EnvoyQuicClientStream::OnStreamReset(const quic::QuicRstStreamFrame& frame)
...
   }
 }
 
 void EnvoyQuicClientStream::ResetWithError(quic::QuicResetStreamError error) {
   ENVOY_STREAM_LOG(debug, "sending reset code={}", *this, error.internal_code());
   stats_.tx_reset_.inc();
   // Upper layers expect calling resetStream() to immediately raise reset callbacks.
-  runResetCallbacks(quicRstErrorToEnvoyLocalResetReason(error.internal_code()));
+  runResetCallbacks(quicRstErrorToEnvoyLocalResetReason(error.internal_code()),
+                    absl::string_view());
```
Comment (alyssawilk) [inline source/common/quic/envoy_quic_client_stream.cc:367]: Is it worth using QuicResetStreamError details here?

File: source/common/quic/envoy_quic_server_stream.cc [hunk 2]
```diff
@@ -353,7 +354,8 @@ void EnvoyQuicServerStream::ResetWithError(quic::QuicResetStreamError error) {
   stats_.tx_reset_.inc();
   if (!local_end_stream_) {
     // Upper layers expect calling resetStream() to immediately raise reset callbacks.
-    runResetCallbacks(quicRstErrorToEnvoyLocalResetReason(error.internal_code()));
+    runResetCallbacks(quicRstErrorToEnvoyLocalResetReason(error.internal_code()),
+                      absl::string_view());
```
Comment (alyssawilk) [inline source/common/quic/envoy_quic_server_stream.cc:358]: ditto

**Review by alyssawilk — APPROVED** [review]
/wait-any happy to merge as-is or wait on a few more details

File: source/common/quic/envoy_quic_client_stream.cc (hunk 1 above)
Comment (RenjieTang) [reply source/common/quic/envoy_quic_client_stream.cc:367]: That's the issue I brought up in the internal bug. If the stream is reset with stream error, it's likely converted to LocalReset in quicRstErrorToEnvoyLocalResetReason and a non-empty transport failure reason will hit the ASSERT at https://github.com/envoyproxy/envoy/blob/0b093461c7ea4fa19b3ad631f048c6673e2b3f1b/source/common/router/upstream_codec_filter.h#L68

Comment (alyssawilk) [reply source/common/quic/envoy_quic_client_stream.cc:367]: right.  my response on the issue was you could remove that assert, rename transport_failure_reason to failure_reason and only set codec_error if failure_reason was empty.  you could do that here or in a follow-up, or just land this as a partial improvement. Thoughts?
/wait-any

Comment (RenjieTang) [reply source/common/quic/envoy_quic_client_stream.cc:367]: Sure I'd like to do that in a follow up

**Pull request merged**